File: src/store/atoms/docker.ts
import { atom } from 'jotai'
import { invoke } from '@tauri-apps/api/core'
import { TauriCommands } from '../../common/tauriCommands'
import { listenEvent, SchaltEvent } from '../../common/eventSystem'
import { logger } from '../../utils/logger'

export interface DockerProjectState {
  available: boolean
  imageExists: boolean
  sandboxEnabled: boolean
  buildInProgress: boolean
}

const dockerProjectStateBaseAtom = atom<DockerProjectState>({
  available: false,
  imageExists: false,
  sandboxEnabled: false,
  buildInProgress: false,
})

export const dockerProjectStateAtom = atom((get) => get(dockerProjectStateBaseAtom))

export const dockerSandboxEnabledAtom = atom(
  (get) => get(dockerProjectStateBaseAtom).sandboxEnabled
)

export const refreshDockerStatusActionAtom = atom(null, async (_get, set) => {
  try {
    const status = await invoke<{
      available: boolean
      imageExists: boolean
      sandboxEnabled: boolean
    }>(TauriCommands.GetDockerStatus)

    set(dockerProjectStateBaseAtom, (prev) => ({
      ...prev,
      available: status.available,
      imageExists: status.imageExists,
      sandboxEnabled: status.sandboxEnabled,
    }))
  } catch (e) {
    logger.error('[Docker] Failed to refresh status', e)
  }
})

export const setDockerSandboxEnabledActionAtom = atom(
  null,
  async (_get, set, enabled: boolean) => {
    try {
      await invoke(TauriCommands.SetDockerSandboxEnabled, { enabled })
      set(dockerProjectStateBaseAtom, (prev) => ({
        ...prev,
        sandboxEnabled: enabled,
      }))
    } catch (e) {
      logger.error('[Docker] Failed to set sandbox enabled', e)
      throw e
    }
  }
)

export const buildDockerImageActionAtom = atom(null, async (_get, set) => {
  set(dockerProjectStateBaseAtom, (prev) => ({ ...prev, buildInProgress: true }))
  try {
    await invoke(TauriCommands.BuildDockerImage)
    set(dockerProjectStateBaseAtom, (prev) => ({
      ...prev,
      imageExists: true,
      buildInProgress: false,
    }))
  } catch (e) {
    set(dockerProjectStateBaseAtom, (prev) => ({ ...prev, buildInProgress: false }))
    logger.error('[Docker] Image build failed', e)
    throw e
  }
})

export const rebuildDockerImageActionAtom = atom(null, async (_get, set) => {
  set(dockerProjectStateBaseAtom, (prev) => ({ ...prev, buildInProgress: true }))
  try {
    await invoke(TauriCommands.RebuildDockerImage)
    set(dockerProjectStateBaseAtom, (prev) => ({
      ...prev,
      imageExists: true,
      buildInProgress: false,
    }))
  } catch (e) {
    set(dockerProjectStateBaseAtom, (prev) => ({ ...prev, buildInProgress: false }))
    logger.error('[Docker] Image rebuild failed', e)
    throw e
  }
})

export const registerDockerEventListenersActionAtom = atom(null, (_get, set) => {
  const unlistenBuild = listenEvent(SchaltEvent.DockerImageBuildProgress, (payload) => {
    if (payload.complete) {
      set(dockerProjectStateBaseAtom, (prev) => ({
        ...prev,
        buildInProgress: false,
        imageExists: payload.success ? true : prev.imageExists,
      }))
    }
  })

  const unlistenStatus = listenEvent(SchaltEvent.DockerStatusChanged, (payload) => {
    set(dockerProjectStateBaseAtom, (prev) => ({
      ...prev,
      available: payload.available,
      imageExists: payload.imageExists,
      sandboxEnabled: payload.sandboxEnabled,
    }))
  })

  return { unlistenBuild, unlistenStatus }
})

File: src-tauri/src/commands/docker.rs
use crate::{DOCKER_MANAGER, PROJECT_MANAGER};
use lucode::domains::docker::service::DockerImageManager;
use lucode::infrastructure::database::db_project_config::ProjectConfigMethods;
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DockerStatus {
    pub available: bool,
    pub image_exists: bool,
    pub sandbox_enabled: bool,
}

#[tauri::command]
pub async fn get_docker_status() -> Result<DockerStatus, String> {
    let available = DockerImageManager::docker_available().await.is_ok();

    let image_exists = if available {
        DockerImageManager::new().image_exists().await
    } else {
        false
    };

    let sandbox_enabled: bool = get_current_project_docker_enabled()
        .await
        .unwrap_or_default();

    Ok(DockerStatus {
        available,
        image_exists,
        sandbox_enabled,
    })
}

#[tauri::command]
pub async fn set_docker_sandbox_enabled(
    app: tauri::AppHandle,
    enabled: bool,
) -> Result<(), String> {
    use crate::events::{DockerStatusChangedPayload, SchaltEvent, emit_event};

    let project = PROJECT_MANAGER
        .get()
        .ok_or_else(|| "Project manager not initialized".to_string())?
        .current_project()
        .await
        .map_err(|e| format!("Failed to get current project: {e}"))?;

    let core = project.schaltwerk_core.write().await;
    let db = core.database();

    db.set_docker_sandbox_enabled(&project.path, enabled)
        .map_err(|e| format!("Failed to set Docker sandbox setting: {e}"))?;

    if !enabled
        && let Some(docker_manager) = DOCKER_MANAGER.get()
        && let Err(e) = docker_manager.stop_container_for_project(&project.path).await
    {
        log::warn!("Failed to stop Docker container after disabling: {e}");
    }

    let available = DockerImageManager::docker_available().await.is_ok();
    let image_exists = if available {
        DockerImageManager::new().image_exists().await
    } else {
        false
    };

    let _ = emit_event(
        &app,
        SchaltEvent::DockerStatusChanged,
        &DockerStatusChangedPayload {
            available,
            image_exists,
            sandbox_enabled: enabled,
        },
    );

    Ok(())
}

async fn run_docker_build(app: &tauri::AppHandle, no_cache: bool) -> Result<(), String> {
    use crate::events::{DockerBuildProgressPayload, SchaltEvent, emit_event};

    let image_manager = DockerImageManager::new();
    let label = if no_cache { "Rebuilding" } else { "Building" };

    let _ = emit_event(
        app,
        SchaltEvent::DockerImageBuildProgress,
        &DockerBuildProgressPayload {
            message: format!("{label} Docker image..."),
            complete: false,
            success: false,
        },
    );

    match image_manager.build_image(no_cache).await {
        Ok(()) => {
            let _ = emit_event(
                app,
                SchaltEvent::DockerImageBuildProgress,
                &DockerBuildProgressPayload {
                    message: "Docker image built successfully".to_string(),
                    complete: true,
                    success: true,
                },
            );
            Ok(())
        }
        Err(e) => {
            let msg = format!("Docker image build failed: {e}");
            let _ = emit_event(
                app,
                SchaltEvent::DockerImageBuildProgress,
                &DockerBuildProgressPayload {
                    message: msg.clone(),
                    complete: true,
                    success: false,
                },
            );
            Err(msg)
        }
    }
}

#[tauri::command]
pub async fn build_docker_image(app: tauri::AppHandle) -> Result<(), String> {
    run_docker_build(&app, false).await
}

#[tauri::command]
pub async fn rebuild_docker_image(app: tauri::AppHandle) -> Result<(), String> {
    run_docker_build(&app, true).await
}

async fn get_current_project_docker_enabled() -> Result<bool, String> {
    let project = PROJECT_MANAGER
        .get()
        .ok_or_else(|| "Project manager not initialized".to_string())?
        .current_project()
        .await
        .map_err(|e| format!("Failed to get current project: {e}"))?;

    let core = project.schaltwerk_core.read().await;
    let db = core.database();

    db.get_docker_sandbox_enabled(&project.path)
        .map_err(|e| format!("Failed to get Docker sandbox setting: {e}"))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_get_docker_status_uninitialized_manager() {
        let result = get_docker_status().await;
        assert!(result.is_ok());
        let status = result.unwrap();
        assert!(!status.sandbox_enabled);
    }
}
